### PR TITLE
Add Dedicated Importer to Viewer

### LINKED ISSUE #30: Programatically validate the form
__Is your feature request related to a problem? Please describe__

In order to provide some feedback to the user by disabling the Task completion button after he/she claims a Task I would like to programmatically validated the form or to be able to retrieve the form errors even though they're not yet visible to the user.

__Describe the solution you'd like__

```js
const form = createForm({
  // ..
});

/*
  Option 1 - Programmatically validating the form:
  Calling this method triggers all validation functions
  and show the errors on the form
*/

form.validate();

// or

/*
  Option 2 - Retrieving the errors:
  The getState() method returns the validation errors
  even if the errors are not yet visible to the user
*/

const {errors} = form.getState();
```

__Describe alternatives you've considered__

Do nothing and just wait for the user to change the form or trigger the submission

__Additional context__

As we talked on this meeting this the reason for these features might not be desirable and I need to sync with the design team, but I'm creating the issue to register the discussion.

### LINKED ISSUE #79: Handle import errors in a robust fashion
__Is your feature request related to a problem? Please describe__

form-js viewer instances need to handle form documents created with different form-js editor versions. To achieve that we've settled down on a feature-based compatibility strategy. This means: Older viewers open form documents as long as they only contain supported features. Newer viewers open older form documents (with the exception of major breaking changes). 

__Describe the solution you'd like__

* [x] The viewer detects if incompatible (unknown to it) features are being used in a form document and fails the import 
* [x] The viewer imports form documents all (success) or nothing (error case)

__Describe alternatives you've considered__

Use the `schemaVersion` to detect compatibility. We did decide to explicitly not go that route. However, the `schemaVersion` as well as the `schemaVersion` baked into the form-js viewer instance can be used to provide a clear error message to the user.

__Additional context__

Part of [versioning strategy](https://github.com/bpmn-io/rfcs/tree/master/004-form-js-versioning), specifically the `form-js@x+2` release.

### PR DESCRIPTION
**Pull request by philippfromme**
* add dedicated importer to viewer
* import each form field adding `id`, `parent` and `path` and adding it to form field registry
* throw error if form field of type not supported or if form field with key already exists
* call `_update` with `field` instead of `path` since `path` is now set at all times
* verify #validate is working

BREAKING CHANGES

* #createForm returns promise resolving with form instead of returning form

Closes #79
Closes #30

Follow-up issues:

* Refactor Editor Import (https://github.com/bpmn-io/form-js/issues/82)

### CODE REVIEW

**Review by nikku — COMMENTED** [review]
Going into a good direction. I inlined some questions.

**Comment by nikku** [comment]
How does this fix #30? Simply by making the import / rendering async? I do see we have a passing test case for it now.

File: packages/form-js-viewer/test/spec/import/Importer.spec.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    try {
+      await form.importSchema(error);
+    } catch (err) {
+
+      // then
+      expect(err).to.exist;
+      expect(err.message).to.eql('form field with key <creditor> already exists');
+    }
+  }));
+
```
Comment (nikku) [inline packages/form-js-viewer/test/spec/import/Importer.spec.js:86]: Can we also add a corrupt JSON error, i.e. import binary or import an incomplete `SELECT` element, i.e. missing options?

**Comment by nikku** [comment]
General proposal: Let's split up adding the validation API and this PR.

From what I understand validation is not a requirement for the new import to work, right?

File: packages/form-js-viewer/src/Form.js [hunk 2]
```diff
@@ -173,12 +173,12 @@ export default class Form {
 
       const fieldErrors = validator.validateField(field, value);
 
-      return set(errors, path, fieldErrors.length ? fieldErrors : undefined);
+      return set(errors, [ pathStringify(path) ], fieldErrors.length ? fieldErrors : undefined);
```
Comment (nikku) [inline packages/form-js-viewer/src/Form.js:176]: How is this related to the validation test?

File: packages/form-js-viewer/src/index.js [hunk 3]
```diff
@@ -34,9 +34,9 @@ export {
  *
  * @param {FormViewerOptions} options
  *
- * @return {Form}
+ * @return {Promise}
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply packages/form-js-viewer/src/index.js]: Done (cf. https://github.com/bpmn-io/form-js/pull/83/files#diff-6e5b8e15b8923a056f4f4c814fb761754b86cac47f0c09bd0a78958eac7f24d8R37 and https://github.com/bpmn-io/form-js/pull/83/commits/9f8f43d6dceab2100f9fc651b5527fd55d2639bd).

File: packages/form-js-viewer/test/spec/Form.spec.js [hunk 4]
```diff
@@ -299,62 +310,4 @@ describe('createForm', function() {
     form.submit();
   });
 
-
-  describe('error handling', function() {
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply packages/form-js-viewer/test/spec/Form.spec.js:303]: Moved to [`Importer.spec.js`](https://github.com/bpmn-io/form-js/pull/83/files#diff-023dda47ee2cfd7020670740515f71a05a093186786bce65036fce5dba6ecd8f).

File: packages/form-js-viewer/test/spec/import/Importer.spec.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  bootstrapForm,
+  getForm,
+  inject
+} from 'test/TestHelper';
+
+import { clone } from 'src/util';
+
+import schema from '../form.json';
+
+describe('Importer', function() {
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply packages/form-js-viewer/test/spec/import/Importer.spec.js:12]: Done (cf. https://github.com/bpmn-io/form-js/pull/83/files#diff-023dda47ee2cfd7020670740515f71a05a093186786bce65036fce5dba6ecd8fR10).

File: packages/form-js-viewer/src/Form.js [hunk 6]
```diff
@@ -104,12 +100,31 @@ export default class Form {
...
-      data: clone(data),
-      errors: {},
-      schema: importSchema(schema).schema
+    this._importedData = null;
+    this._importedSchema = null;
+
+    return new Promise((resolve, reject) => {
+      const importer = this.get('importer');
+
+      schema = clone(schema);
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply packages/form-js-viewer/src/Form.js:108]: Actually, we only need to clone data twice. First, we clone it to avoid mutating the `data` that was passed and then we clone it again to keep a copy of `data` for resetting the form. I cleaned it up (cf. https://github.com/bpmn-io/form-js/pull/83/files#diff-53d9b079c7db646b79736d7ab07cd7d3108049b780d14637f41c188a4f6ffc03R105).

File: packages/form-js-viewer/src/Form.js [hunk 7]
```diff
@@ -104,12 +100,31 @@ export default class Form {
...
+          this._setState({
+            data,
+            errors: {},
+            schema
+          });
+
+          resolve({ warnings });
+        })
+        .catch(err => {
+          reject(err);
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply packages/form-js-viewer/src/Form.js:124]: Warnings are exposed through the error received from the importer (cf. https://github.com/bpmn-io/form-js/pull/83/files#diff-9f0268e4ce241c1b7c182fce6fff8d93a27c9a5db106c4366eb99def13de44bbR33).

**Comment by philippfromme** [comment]
> How does this fix #30? Simply by making the import / rendering async? I do see we have a passing test case for it now.

Validating the form right after creating it using `createForm` wasn't previously possible due to the fact that form fields were added to the form field registry only when they were rendered. Therefore, the form field registry wouldn't contain all form fields right after creating the form and thus validating all form fields wasn't possible. It is now.

File: packages/form-js-viewer/test/spec/import/Importer.spec.js (hunk 1 above)
Comment (philippfromme) [reply packages/form-js-viewer/test/spec/import/Importer.spec.js:135]: I've added 2 tests for both broken JSON and incomplete form fields (cf. https://github.com/bpmn-io/form-js/pull/83/files#diff-023dda47ee2cfd7020670740515f71a05a093186786bce65036fce5dba6ecd8fR94). The error handling can be improved. Furthermore, we cannot catch errors due to incomplete form fields at the moment as they will only occur during rendering.

File: packages/form-js-viewer/src/Form.js (hunk 2 above)
Comment (philippfromme) [reply packages/form-js-viewer/src/Form.js:176]: The reason using `path` instead of `[ pathStringify(path) ]` works is that we don't support nesting at the moment. Once we add support for nesting it would break as it would create nested objects instead of something like `documents.0.author: [ ... ]` as defined in the `@typedef` of `Errors`.

**Comment by philippfromme** [comment]
> General proposal: Let's split up adding the validation API and this PR.
>
> From what I understand validation is not a requirement for the new import to work, right?

Validation is not a requirement. It's something that is fixed by the refactored import.

**Review by nikku — APPROVED** [review]
Looks good.

**Comment by nikku** [comment]
One thing missing though. We do test error handling via internals (`Importer.spec.js`). We do not test error handling via the public API `createForm` anymore. We should do that.
